### DISCUSSION TITLE
Using python coverage from an avocado run

### DISCUSSION THREAD

**Discussion by pevogam** [avocado-framework/avocado#5610] [Q&A] [answered]
Hi all, I am running some unit tests for our plugin and use avocado for all of them. However, I generally also submit coverage reports which worked easily with unit tests but no longer seem to be effective with avocado. What I run is
```
PYTHONPATH=:. coverage run --append --source=avocado_i2n $(which avocado) run --nrunner-max-parallel-tasks=4 selftests/isolation/test_*
```
but the resulting coverage is 0% throughout. Is there a standard way to invoke avocado together with python3-coverage? Would you suggest a better way then the one above?

**Comment by arif-ali** [comment]
I am also keen to understand this too, as per the [documentation](https://avocado-framework.readthedocs.io/en/latest/guides/writer/chapters/integrating.html) it suggests that we should get coverage via invoking coverage and then pass on avocado but I get no-data-collected

```
<snip>/lib/python3.12/site-packages/coverage/control.py:887: CoverageWarning: No data was collected. (no-data-collected)
  self._warn("No data was collected.", slug="no-data-collected")
```

TiA

**Reply by richtja** [reply]
Hi @arif-ali, I have investigated this, and I can confirm that we have some issues here and the coverage doesn't work as is written in our documentation. Therefore, I have created #5919 to track this issue and I will be working on a solution.

I have found out that if you will set env variable `COVERAGE_RUN` before testing, the coverage will work as expected, see [this](https://github.com/avocado-framework/avocado/blob/master/avocado/plugins/runners/avocado_instrumented.py#L101). I hope this will help you until I fix it and put everything into a good shape.

**Reply by arif-ali** [reply]
@richtja I had a go at this, but didn't have much luck.

I've tried in the `tox` environment, manually and even tried avocado directly with the env variable, but didn't have much luck. Below is a paste of the output of the current run, which shows the env, and the coverage runs.

https://dpaste.com/F9R86XMZH

**Reply by richtja** [reply]
Hi @arif-ali can you please also provide your `avocado config` and `.coveragerc` file. I will investigate it. Thank you

**Reply by arif-ali** [reply]
Hi @richtja below are the requested details

* `avocado config`: https://dpaste.com/CXRWYXUDM
* `.coveragerc`: https://dpaste.com/BN6UABPES

**Comment by richtja** [comment, marked as answer]
Hi @pevogam and @arif-ali the #5944 should solve the issues with the coverage.py on Avocado. Can you please try it out and let me know if it works for you? Thank you.

**Reply by arif-ali** [reply]
Seems to have worked for me https://paste.ubuntu.com/p/M6vh6Bsjzy/, I made the following change in our `tox.ini`  to run from the wheel I got out of the code

```diff
diff --git a/tox.ini b/tox.ini
index e1952613..91990328 100644
--- a/tox.ini
+++ b/tox.ini
@@ -51,10 +51,17 @@ basepython = python3
 commands =
     {[testenv]avocado_cmd} -t foreman {posargs:{[testenv]foreman_tests}}
 
-[testenv:nosetests]
+[testenv:coverage]
 basepython = python3
+setenv =
+    {[testenv]setenv}
+    COVERAGE_RUN = "true"
 deps =
-    {[testenv]deps}
-    nose3
+    #{[testenv]deps}
+    ./avocado_framework-105.0-py2.py3-none-any.whl
+    coverage
 commands =
-    nosetests -v --with-coverage --cover-package=sos tests/unittests --cover-html
+    coverage erase
+    coverage run -m avocado run tests/unittests
+    coverage combine
+    coverage report
```

**Reply by richtja** [reply]
Great, thank you for trying it out. I will work on back port to 103lts

**Reply by pevogam** [reply]
Hi @richtja so the original line I used (as can be seen above) was:
```
# in ci.yml
run: PYTHONPATH=:. coverage run --append --source=avocado_i2n $(which avocado) run --max-parallel-tasks=4 selftests/isolation/test_*
```
and I see that your fix involves exporting an environmental variable like `COVERAGE_RUN`, then a file with the sources like `.coveragerc` and then running avocado directly with new coverage>=7.5. But the above line still did not work with
```
 Collecting coverage (from -r requirements_pip.txt (line 1))
  Using cached coverage-7.5.3-cp311-cp311-manylinux_2_5_x86_64.manylinux1_x86_64.manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata (8.2 kB)
```
and doesn't need the extra rc file to add the sources as it does so directly. My suspicion is it also doesn't need the env variable as it does not invoke avocado directly. Despite all of this in the end I still see
```
 /opt/hostedtoolcache/Python/3.11.9/x64/lib/python3.11/site-packages/coverage/control.py:888: CoverageWarning: No data was collected. (no-data-collected)
  self._warn("No data was collected.", slug="no-data-collected")
```
Is the direct avocado invocation the only one we are allowed to use?

**Reply by arif-ali** [reply]
Hi @pevogam I see the same message, but it actually does do the coverage

```
/home/arif/gitRepos/sos/.tox/coverage/lib/python3.12/site-packages/coverage/control.py:888: CoverageWarning: No data was collected. (no-data-collected)
  self._warn("No data was collected.", slug="no-data-collected")
```

You need to run the following commands afterwards to get the report, and you do need the COVERAGE_RUN env variable, as I found, it doesn't work without that
```
coverage combine
coverage report
```

I have also found that either of the following would work for me with the env set

```
    avocado run {posargs:{[testenv]unit_tests}}
```
or
```
    coverage run -m avocado run {posargs:{[testenv]unit_tests}}
```

**Reply by richtja** [reply]
Hi @pevogam, yes, you are right `COVERAGE_RUN` doesn't need to be set, because the `coverage.py` sets this variable by itself. But the  `.coveragerc` is necessary, because of avocados test isolation. Avocado is using a coverage API on the test runner's side and unfortunately the API has only access to the `.coveragerc` and not to the directly set configuration. If you want to get more information about how it is done, you can check the [avocado-instrumented implementation](https://github.com/avocado-framework/avocado/blob/master/avocado/plugins/runners/avocado_instrumented.py#L107). 

Also, please check if your coverage is installed with[ C Extension](https://coverage.readthedocs.io/en/7.5.3/install.html#checking-the-installation).

Even tho you will set everything correctly, you can still get `No data was collected.` warning, because the initial `coverage run` hasn't collected any data, because the data have been collected by coverage API on the test runner's side. Therefore, you should still be able to run `coverage combine` and `coverage report` to get the coverage.

**Reply by arif-ali** [reply]
yup, my bad, `COVERAGE_RUN` does not need to be set; just re-tested with the `coverage` command again

**Reply by pevogam** [reply]
Alright then, indeed it could be made to work right away this way give or take the `No data was collected.` which may confuse unknowing users. I am still a bit concern about the big difference in the running time I obtained though - a run with avocado (4 workers) without coverage will take [~8 minutes](https://github.com/intra2net/avocado-i2n/actions/runs/9478206573):
![image](https://github.com/avocado-framework/avocado/assets/1969979/9d8b011e-2158-4fed-8528-88eadbd4a41d)
while a run with avocado and coverage will instead take [~30 minutes](https://github.com/intra2net/avocado-i2n/actions/runs/9485331957):
![image](https://github.com/avocado-framework/avocado/assets/1969979/fb94c3e1-767f-46ac-8792-ddfbd7d23d9e)
which is a bit surprising even considering that coverage could take longer to count all calls and code. I could of course compare this with coverage from standard python `unittest` but it is running in serial mode so a comparison is not exactly useful.

Perhaps we need additional setting for parallel runs?

**Reply by richtja** [reply]
@pevogam, thanks for running this test. 

> Alright then, indeed it could be made to work right away this way give or take the `No data was collected.` which may confuse unknowing users.

Yes, you are right, I will update our documentation to inform users about this.

> I am still a bit concern about the big difference in the running time I obtained though - a run with avocado (4 workers) without coverage will take [~8 minutes](https://github.com/intra2net/avocado-i2n/actions/runs/9478206573): ![image](https://private-user-images.githubusercontent.com/1969979/339046294-9d8b011e-2158-4fed-8528-88eadbd4a41d.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTgyNjQ2NTQsIm5iZiI6MTcxODI2NDM1NCwicGF0aCI6Ii8xOTY5OTc5LzMzOTA0NjI5NC05ZDhiMDExZS0yMTU4LTRmZWQtODUyOC04OGVhZGJkNGE0MWQucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDYxMyUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA2MTNUMDczOTE0WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9OGVlNjYyNjA0OTgzZWIzZjYxOWZjZWJhOWQyNzMxN2E0MDA2MzdmYmJhNzQ2MmM3NTJhOWViYzc3MGM5ZTI1OCZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.mriEpalCYDkcSAXv3e2-hWHDQ6XbgGlTV-Fc-TII57c) while a run with avocado and coverage will instead take [~30 minutes](https://github.com/intra2net/avocado-i2n/actions/runs/9485331957): ![image](https://private-user-images.githubusercontent.com/1969979/339046620-fb94c3e1-767f-46ac-8792-ddfbd7d23d9e.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTgyNjQ2NTQsIm5iZiI6MTcxODI2NDM1NCwicGF0aCI6Ii8xOTY5OTc5LzMzOTA0NjYyMC1mYjk0YzNlMS03NjdmLTQ2YWMtODc5Mi1kZGZiZDdkMjNkOWUucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDYxMyUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA2MTNUMDczOTE0WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9NGNiMDAzZGZiM2VmNDdhMjY0YjEzZjUxN2E0N2M2OWI1NmEyMzdmMDljZjgyMWI4OTJjMWVjYzExNDUwMzkzMCZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.fS79m8SnzURH_yxk2RwAqlhD6kTmXXUKmGRXsmejSvY) which is a bit surprising even considering that coverage could take longer to count all calls and code. I could of course compare this with coverage from standard python `unittest` but it is running in serial mode so a comparison is not exactly useful.
> 
> Perhaps we need additional setting for parallel runs?

Can you please compare it with the coverage on standard python `unittest`, because I don't see anything like this on avocado selftests. The [coverage test](https://github.com/avocado-framework/avocado/actions/runs/9320516334/job/25657474817?pr=5944) run for 9m 17s and the [regular test](https://github.com/avocado-framework/avocado/actions/runs/9320516334/job/25657557577?pr=5944) run for 7m 39s.

**Reply by pevogam** [reply]
Unfortunately as I wrote above I don't think a comparison with a `unittest` variant is possible despite having an [executed workflow](https://github.com/intra2net/avocado-i2n/actions/runs/9485740406) to compare with since it runs serially while I use avocado with `--max-parallel-tasks=4`. From the comparison above it seems like the coverage job is ~4 times slower, could it be that the parallel run is not really possible and the tasks end up waiting for each other?

**Reply by pevogam** [reply]
I did read somewhere in the diff about previous configuration that needed special `parallel` handling with coverage. In any case perhaps I am wrong and looking at the variants separately it seems the time for each has more or less doubled. So it could also be some result of variation in the servers load, I simply see that form ~8 mins per run before now multiple runs are at an average of ~25-26 minutes.

**Reply by richtja** [reply]
The parallel handling is not needed from the point of avocado. Avocado will run separated coverage.py process for each test and the results will be stored into separated file, therefore you have to run `coverage combine` after the testing to merge the results into one report. On the other hand, if one of your tests is concurrent, then you need to set [concurrency](https://coverage.readthedocs.io/en/7.5.3/config.html#run-concurrency) value in `.covaregerc` file. Unfortunately, IIUIC this shouldn't have any impact on the runtime, only on the coverage report.
